src/processing.py

```diff
@@ -1,12 +1,85 @@
 from pathlib import Path
 import time
+import paramiko
+import os
+import base64
+import tempfile
+import shutil
 
-from .supabase import use_client, login
+from fastapi import HTTPException
+
+from .supabase import use_client, login, verify_token
 from .settings import settings
-from .models import StatusEnum, Dataset, QueueTask, Cog
+from .models import StatusEnum, Dataset, QueueTask, Cog, Thumbnail
 from .logger import logger
 from . import monitoring
 from .deadwood.cog import calculate_cog
+from .deadwood.thumbnail import calculate_thumbnail
+
+
+def pull_file_from_storage_server(remote_file_path: str, local_file_path: str):
+    # Check if the file already exists locally
+    if os.path.exists(local_file_path):
+        print(f"File already exists locally at: {local_file_path}")
+        return
+
+    ssh = paramiko.SSHClient()
+    ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+    print(
+        "connecting to:",
+        settings.storage_server_ip,
+        settings.storage_server_username,
+        settings.storage_server_password,
+    )
+    ssh.connect(
+        hostname=settings.storage_server_ip,
+        username=settings.storage_server_username,
+        password=settings.storage_server_password,
+        port=22,  # Add this line to specify the default SSH port
+    )
+
+    sftp = ssh.open_sftp()
+    print("pulling file: ", remote_file_path, "to", local_file_path)
+
+    # Create the directory for local_file_path if it doesn't exist
+    local_dir = Path(local_file_path).parent
+    local_dir.mkdir(parents=True, exist_ok=True)
+
+    sftp.get(remote_file_path, local_file_path)
+    print("file pulled")
+    sftp.close()
+    ssh.close()
+
+    # Check if the file exists after pulling
+    if os.path.exists(local_file_path):
+        print(f"File successfully saved at: {local_file_path}")
+        print(f"File size: {os.path.getsize(local_file_path)} bytes")
+    else:
+        print(f"Error: File not found at {local_file_path} after pulling")
+
+
+def push_file_to_storage_server(local_file_path: str, remote_file_path: str):
+    ssh = paramiko.SSHClient()
+    ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+    ssh.connect(
+        hostname=settings.storage_server_ip,
+        username=settings.storage_server_username,
+        password=settings.storage_server_password,
+        port=22,  # Add this line to specify the default SSH port
+    )
+    sftp = ssh.open_sftp()
+    # Extract the remote directory path
+    remote_dir = os.path.dirname(remote_file_path)
+
+    try:
+        # Create the remote directory if it doesn't exist
+        sftp.mkdir(remote_dir)
+    except IOError:
+        # Directory might already exist, which is fine
+        pass
+    sftp.put(local_file_path, remote_file_path)
```

Comment:
Can we test what happens when the connection is cut? Would it restart or just fail? May want to have try except redo loop around this entire function.

Reply:
yes, this is not really implemented yet. But we need a mechanism for the task to rerun in case of failure. This is not the case at the moment. 
There are several things that can go wrong (server restart, crash of process, docker not being available ..)
I will address this as sooon as we can experiment with the entire system. Since this is difficult to simulate.